fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBSystemOperations.java

```diff
@@ -0,0 +1,162 @@
+     */
+    @Nonnull
+    public static CompletableFuture<String> getPrimaryDatacenterAsync(@Nonnull FDBDatabaseRunner runner) {
+        return runner.runAsync(context -> {
+            final Transaction tr = context.ensureActive();
+            tr.options().setReadSystemKeys();
+            return tr.get(SystemKeyspace.PRIMARY_DATACENTER_KEY).thenApply(FDBSystemOperations::nullableUtf8);
+        });
+    }
+
+    /**
+     * Get the primary datacenter of the underlying cluster. This is a blocking version of
+     * {@link #getPrimaryDatacenterAsync(FDBDatabaseRunner)}. If the FDB cluster's primary datacenter has not
+     * been set, this will return {@code null}.
+     *
+     * @param runner a runner to use to perform the operation
+     * @return the primary datacenter of the database underlying {@code runner}
+     * @see #getPrimaryDatacenterAsync(FDBDatabaseRunner)
+     */
+    @Nullable
+    public static String getPrimaryDatacenter(@Nonnull FDBDatabaseRunner runner) {
+        return asyncToSync(runner, getPrimaryDatacenterAsync(runner));
+    }
+
+    @Nonnull
+    private static CompletableFuture<String> getConnectionStringAsyncInternal(@Nonnull FDBRecordContext context) {
+        return context.ensureActive().get(SystemKeyspace.CONNECTION_STR_KEY).thenApply(FDBSystemOperations::nullableUtf8);
+    }
+
+    /**
+     * Get the connection string used to connect to the FDB cluster. This string essentially contains the contents of
+     * the cluster file, though it may change if the cluster's coordinators change. Note that even though this
+     * operation requires having a database connection (in the transaction) and returns a future (as the client
+     * may schedule to complete the work at a later time), it does not actually perform any network calls but reads
+     * the value from the client's local memory.
+     *
+     * <p>
+     * For more information, see the documentation on the
+     * <a href="https://apple.github.io/foundationdb/administration.html#cluster-file-format">cluster file format</a>
+     * in the FoundationDB documentation.
+     * </p>
+     *
+     * @param runner a runner to use to perform the operation
+     * @return a future that will contain the current cluster connection string
+     */
+    @Nonnull
+    public static CompletableFuture<String> getConnectionStringAsync(@Nonnull FDBDatabaseRunner runner) {
+        return runner.runAsync(FDBSystemOperations::getConnectionStringAsyncInternal);
+    }
+
+    /**
+     * Get the connection string used to connect to the FDB cluster. This is a synchronous version of
+     * {@link #getConnectionStringAsync(FDBDatabaseRunner)}
+     *
+     * @param runner a runner to use to perform the operation
+     * @return the current cluster connection string
+     * @see #getConnectionStringAsync(FDBDatabaseRunner)
+     * @see #getConnectionStringAsync(FDBDatabaseRunner)
```

Comment:
```suggestion
```

Reply:
Whoops, must have added the dupe at some point